src/relay/op/tensor/reduce.cc

```diff
@@ -0,0 +1,210 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file reduce.cc
+ * \brief Reduction operators.
+ */
+#include <tvm/relay/expr.h>
+#include <tvm/relay/op.h>
+#include "../type_relations.h"
+
+namespace tvm {
+namespace relay {
+
+/*! \brief Attributes for Reduce operators */
+struct ReduceAttrs : public tvm::AttrsNode<ReduceAttrs> {
+  Array<IndexExpr> axis;
+  bool keepdims;
+  bool exclude;
+
+  TVM_DECLARE_ATTRS(ReduceAttrs, "relay.attrs.ReduceAttrs") {
+    TVM_ATTR_FIELD(axis).set_default(Array<IndexExpr>({}))
+        .describe(R"code(The axis or axes along which to perform the reduction.
+
+      The default, `axis=()`, will compute over all elements into a
+      scalar array with shape `(1,)`.
+
+      If `axis` is int, a reduction is performed on a particular axis.
+
+      If `axis` is a tuple of ints, a reduction is performed on all the axes
+      specified in the tuple.
+
+      If `exclude` is true, reduction will be performed on the axes that are
+      NOT in axis instead.)code");
+
+    TVM_ATTR_FIELD(keepdims).set_default(false)
+      .describe("If this is set to `True`, the reduced axes are left "
+                "in the result as dimension with size one.");
+    TVM_ATTR_FIELD(exclude).set_default(false)
+      .describe("Whether to perform reduction on axis that are NOT in axis instead.");
+  }
+};
+
+inline std::vector<IndexExpr> GetReduceAxes(const uint32_t indim,
+                                            const std::vector<IndexExpr>& axis,
+                                            bool exclude,
+                                            const TypeReporter& reporter) {
+  if (axis.size() == 0) {
+    std::vector<IndexExpr> r_axes;
+    r_axes.push_back(make_const(tvm::Int(64), 0));
+    return r_axes;
+  }
+  auto in_dim = make_const(tvm::Int(64), indim);
+  CHECK(reporter->Assert(axis[axis.size() - 1] < in_dim))
+    << "Reduction axis " << axis[axis.size() - 1]
+    << " exceeds input dimensions " << indim;
+
+  std::vector<IndexExpr> in_axis = axis;
```

Comment:
Unfortunately, this code was written as if it supports symbolic case, but really requires constant integer. In this case, we should simply require the axis to be all concrete integers, in this case, use as_const_int to get the integer pointer, assert that pointer is not null, and get value out.

Reply:
store things as vector<int64_t>